compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToUKernels.cpp

```diff
@@ -0,0 +1,271 @@
+/// ================== SAME UTILITIES AS IREE LLVMCPU ====================
+/// ======================================================================
+
+/// Returns `true` if an `outsOperand` value is initialized to zero.
+static bool isInitializedToZero(Value outsOperand) {
+  auto fillOp = outsOperand.getDefiningOp<linalg::FillOp>();
+  if (!fillOp) return false;
+  Value fillVal = fillOp.getDpsInputOperand(0)->get();
+  return matchPattern(fillVal, m_Zero()) ||
+         matchPattern(fillVal, m_AnyZeroFloat());
+}
+
+/// Holds a function name and attributes.
+struct FnNameAndDefAttrs {
+  std::string name;
+  SmallVector<NamedAttribute> defAttrs;
+};
+
+/// Returns the function name and attributes to use for a ukernel with given
+/// `ukernelName` on the target described by `targetAttr`.
+static FnNameAndDefAttrs getFnNameAndDefAttrs(AIEPassPipeline passPipeline,
+                                              std::string ukernelName,
+                                              std::string inputOutputElemType) {
+  FnNameAndDefAttrs result;
+  std::string ukernelSuffix = "";
+  if (passPipeline == AIEPassPipeline::PadPipeline) ukernelSuffix = "_scalar";
+  result.name = ukernelName + ukernelSuffix + "_" + inputOutputElemType;
+  return result;
+}
+
+/// ============================= END ====================================
+/// ================== SAME UTILITIES AS IREE LLVMCPU ====================
+/// ======================================================================
+
+/// TODO(avarma): This currently is skipping checking for ext* ops.
+/// TODO(avarma): Will shift this utility to a common space later to be used by
+///               KernelDispatch as well.
+static bool bodyMatcherForMatmul(Value yieldVal, Block *body) {
+  Operation *addOp = yieldVal.getDefiningOp();
+  if (!isa_and_nonnull<arith::AddIOp, arith::AddFOp>(addOp)) {
+    return false;
+  }
+  Operation *mulOp = addOp->getOperand(1).getDefiningOp();
+  if (!isa_and_nonnull<arith::MulIOp, arith::MulFOp>(mulOp)) {
+    return false;
+  }
+  auto lhsBlockArg = mulOp->getOperand(0).dyn_cast<BlockArgument>();
+  auto rhsBlockArg = mulOp->getOperand(1).dyn_cast<BlockArgument>();
+  auto outBlockArg = addOp->getOperand(0).dyn_cast<BlockArgument>();
+  if (!lhsBlockArg || !rhsBlockArg || !outBlockArg ||
+      lhsBlockArg.getOwner() != body || rhsBlockArg.getOwner() != body ||
+      outBlockArg.getOwner() != body || lhsBlockArg.getArgNumber() != 0 ||
+      rhsBlockArg.getArgNumber() != 1 || outBlockArg.getArgNumber() != 2) {
+    return false;
+  }
+  return true;
+}
+
+/// `isMatmul` is a utility function that aims to indentify whether a
+/// linalg op is a matmul op.
+static bool isMatmul(linalg::LinalgOp linalgOp) {
+  // Step 0. Test if the op itself is a linalg.matmul op.
+  if (isa<linalg::MatmulOp>(linalgOp)) return true;
+  // Step 1. Test the body of the generic to indeed be what we expect for a
+  //         matmul.
```

Comment:
Sure, I can add that as another `if` following right after the `isa<linalg::MatmulOp>(op)` check.

Although I think that wouldn't be necessary though since this part of the code is currently invoked by `linalg.generic/matmul` instances only - since we might be extending this for other "matmul" variants and anyway have to let go of this check.